### PR TITLE
Add method to iterate through a DEM which decomposes Hyper Edges

### PR DESCRIPTION
**Pull request by smadhuk**
This PR introduces `iter_dem_instructions_include_correlations`, a helper method that iterates through `error` instructions in the DEM and processes components of a suggested decomposition (with fewer than 3 detectors) of a hyper edge as per a user-provided handler. In a future PR, this method will be used to capture information about correlations between decomposed error mechanisms.

It also re-adds in tests for `merge_weights` that were removed erroneously in a prior change.

We also establish a baseline for loading a DEM with correlated information included:

```
[.................*<<|....................] 530 us (vs 290 us) (2.8 Mdets/s) (9.5 Mlayers/s) (1.9 Mshots/s) Decode_surface_r5_d5_p1000
[................*<<<|....................]  23 ms (vs  10 ms) (920 kdets/s) ( 59 klayers/s) (5.3 kshots/s) Decode_surface_r11_d11_p100
[................*<<<|....................] 3.4 ms (vs 1.5 ms) (2.9 Mdets/s) (1.6 Mlayers/s) (140 kshots/s) Decode_surface_r11_d11_p1000
[..................*<|....................] 140 us (vs  83 us) (3.6 Mdets/s) ( 19 Mlayers/s) (1.7 Mshots/s) Decode_surface_r11_d11_p10000
[..................*<|....................]  56 us (vs  33 us) (3.3 Mdets/s) (190 Mlayers/s) ( 18 Mshots/s) Decode_surface_r11_d11_p100000
[................*<<<|....................]  16 ms (vs 7.5 ms) (600 kdets/s) (9.9 klayers/s) (470 shots/s) Decode_surface_r21_d21_p100
[...............*<<<<|....................]  23 ms (vs 7.8 ms) (430 kdets/s) (7.1 klayers/s) (330 shots/s) Decode_surface_r21_d21_p100_with_dijkstra
[...............*<<<<|....................]  24 ms (vs 8.2 ms) (410 kdets/s) (6.8 klayers/s) (320 shots/s) Decode_surface_r21_d21_p100_to_edges
[.................*<<|....................]  13 ms (vs 6.3 ms) (2.7 Mdets/s) (390 klayers/s) ( 18 kshots/s) Decode_surface_r21_d21_p1000
[..............*<<<<<|....................]  28 ms (vs 7.7 ms) (1.2 Mdets/s) (180 klayers/s) (8.9 kshots/s) Decode_surface_r21_d21_p1000_with_dijkstra
[..............*<<<<<|....................]  30 ms (vs 8.4 ms) (1.1 Mdets/s) (170 klayers/s) (8.2 kshots/s) Decode_surface_r21_d21_p1000_to_edges
[................*<<<|....................] 2.4 ms (vs 980 us) (3.2 Mdets/s) (4.4 Mlayers/s) (210 kshots/s) Decode_surface_r21_d21_p10000
[..............*<<<<<|....................] 5.0 ms (vs 1.3 ms) (1.5 Mdets/s) (2.1 Mlayers/s) (100 kshots/s) Decode_surface_r21_d21_p10000_with_dijkstra
[..............*<<<<<|....................] 5.4 ms (vs 1.4 ms) (1.4 Mdets/s) (1.9 Mlayers/s) ( 94 kshots/s) Decode_surface_r21_d21_p10000_to_edges
[.................*<<|....................] 180 us (vs  94 us) (3.9 Mdets/s) ( 58 Mlayers/s) (2.7 Mshots/s) Decode_surface_r21_d21_p100000
[...............*<<<<|....................] 380 us (vs 130 us) (1.8 Mdets/s) ( 27 Mlayers/s) (1.3 Mshots/s) Decode_surface_r21_d21_p100000_with_dijkstra
[...............*<<<<|....................] 430 us (vs 130 us) (1.6 Mdets/s) ( 24 Mlayers/s) (1.1 Mshots/s) Decode_surface_r21_d21_p100000_to_edges
[..................*<|....................]  54 ms (vs  35 ms) (180 loads/s) Load_dem_r11_d11_p100
[..................*<|....................] 450 ms (vs 280 ms) ( 22 loads/s) Load_dem_r21_d21_p100
[....................|>*..................]  24 ms (vs  35 ms) (410 loads/s) Load_dem_r11_d11_p100_correlations
[....................|>*..................] 170 ms (vs 280 ms) ( 57 loads/s) Load_dem_r21_d21_p100_correlations
[................*<<<|....................] 9.7 ms (vs 3.5 ms) (1.0 Gcalls/s) Varying32_get_distance_at_time
[..............*<<<<<|....................]  22 ms (vs 5.9 ms) (450 Mcalls/s) Varying64_get_distance_at_time
[....................*....................] 4.8 us (vs 4.9 us) (200 MEnqueueDequeues/s) bucket_queue_sort
[..................*<|....................] 140 us (vs  99 us) ( 67 MEnqueueDequeues/s) bucket_queue_stream
```

### CODE REVIEW

File: src/pymatching/sparse_blossom/driver/user_graph.h [hunk 1]
```diff
@@ -228,6 +228,72 @@ void iter_detector_error_model_edges(
...
+        double p = instruction.arg_data[0];
+        pm::DecomposedDemError decomposed_err;
+        decomposed_err.probability = p;
+        decomposed_err.components = {};
+        decomposed_err.components.push_back({});
+        UserEdge* component = &decomposed_err.components.back();
+        // Mark component as empty to begin with.
+        component->node1 = SIZE_MAX;
+        component->node2 = SIZE_MAX;
+        size_t num_translated_relative_detectors = 0;
```
Comment (oscarhiggott) [inline src/pymatching/sparse_blossom/driver/user_graph.h:255]: Maybe `num_component_detectors` instead of `num_translated_relative_detectors` would be a better name?

File: src/pymatching/sparse_blossom/driver/user_graph.h [hunk 2]
```diff
@@ -228,6 +228,72 @@ void iter_detector_error_model_edges(
...
+                    // We mark errors which have 3 or more detectors as a special boundary-to-boundary edge.
+                    component->node1 = SIZE_MAX;
+                    component->node2 = SIZE_MAX;
+                }
+            } else if (target.is_observable_id()) {
+                component->observable_indices.push_back(target.val());
+            } else if (target.is_separator()) {
+                // If the previous error in the decomposition had 3 or more components, we ignore it.
+                if (component->node1 == SIZE_MAX) {
+                    decomposed_err.components.pop_back();
```
Comment (oscarhiggott) [inline src/pymatching/sparse_blossom/driver/user_graph.h:275]: What you have here (ignoring the hyperedge) is fine, but it would also be ok to throw an exception here. Since we know there is a separator (e.g. this error is decomposed), we would really expect each component to be edge-like.

File: src/pymatching/sparse_blossom/driver/user_graph.test.cc [hunk 3]
```diff
@@ -166,3 +181,144 @@ TEST(UserGraph, DecodeUserGraphDetectionEventOnBoundaryNode) {
...
+    ASSERT_EQ(handler.handled_errors.size(), 3);
+
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.1, 0, 1, {}}));
+    EXPECT_EQ(handler.handled_errors[1], (HandledError{0.1, 2, 3, {0}}));
+    EXPECT_EQ(handler.handled_errors[2], (HandledError{0.1, 4, SIZE_MAX, {}}));
+}
+
+// Test a decomposed error where one of the components is a hyperedge and should be ignored.
+TEST(IterDemInstructionsTest, DecomposedErrorWithIgnoredComponent) {
+    stim::DetectorErrorModel dem("error(0.15) D0 D1 ^ D2 D3 D4 ^ D5 D6 L2");
```
Comment (oscarhiggott) [inline src/pymatching/sparse_blossom/driver/user_graph.test.cc:287]: As mentioned above, this is fine but might be slightly better to throw an exception if an error is decomposed (contains at least one `^`) but has a component that is a hyperedge.

File: src/pymatching/sparse_blossom/driver/user_graph.test.cc [hunk 4]
```diff
@@ -166,3 +181,144 @@ TEST(UserGraph, DecodeUserGraphDetectionEventOnBoundaryNode) {
...
+    )DEM");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+
+    // We expect 5 errors to be handled from this model.
+    // 1. {0.1, 0, MAX, {}} from "error(0.1) D0"
+    // 2. {0.2, 1, 2, {0}} from "error(0.2) D1 D2 L0"
+    // 3. {0.3, 6, MAX, {}} from the second part of "error(0.3) D3 D4 D5 ^ D6" (first part is ignored)
+    // 4. {0.4, 8, MAX, {}} from the first part of "error(0.4) D8 ^ D9 L1"
+    // 5. {0.4, 9, MAX, {1}} from the second part of "error(0.4) D8 ^ D9 L1"
```
Comment (oscarhiggott) [inline src/pymatching/sparse_blossom/driver/user_graph.test.cc:313]: Could remove these commented lines?

File: src/pymatching/sparse_blossom/driver/user_graph.test.cc [hunk 5]
```diff
@@ -166,3 +181,144 @@ TEST(UserGraph, DecodeUserGraphDetectionEventOnBoundaryNode) {
...
+    std::vector<HandledError> expected = {{0.15, 0, 1, {}}, {0.15, 5, 6, {2}}};
+    EXPECT_EQ(handler.handled_errors, expected);
+}
+
+// Test a complex DEM with multiple instruction types and edge cases combined.
+TEST(IterDemInstructionsTest, CombinedComplexDem) {
+    stim::DetectorErrorModel dem(R"DEM(
+        error(0.1) D0            # Instruction 1: Simple
+        error(0.2) D1 D2 L0      # Instruction 2: Two detectors, one observable
+        error(0.3) D3 D4 D5 ^ D6 # Instruction 3: Hyperedge ignored, second component handled
```
Comment (oscarhiggott) [inline src/pymatching/sparse_blossom/driver/user_graph.test.cc:301]: Could replace this with a single hyperedge:

`error(0.3) D3 D4 D5`
 
since I think it is fine to ignore this rather than throwing an exception.

**Review by oscarhiggott — COMMENTED** [review]
Looks great! Just added some minor suggestions.

File: src/pymatching/sparse_blossom/driver/user_graph.h (hunk 1 above)
Comment (smadhuk) [reply src/pymatching/sparse_blossom/driver/user_graph.h]: Done

File: src/pymatching/sparse_blossom/driver/user_graph.h (hunk 2 above)
Comment (smadhuk) [reply src/pymatching/sparse_blossom/driver/user_graph.h:275]: The suggestion makes sense; will do in a follow up since I have a few PRs coming in that are based off of this branch and it will be cleaner.

File: src/pymatching/sparse_blossom/driver/user_graph.test.cc (hunk 3 above)
Comment (smadhuk) [reply src/pymatching/sparse_blossom/driver/user_graph.test.cc:287]: The suggestion makes sense; will do in a follow up since I have a few PRs coming in that are based off of this branch and it will be cleaner
